### PR TITLE
Read last N messages function

### PR DESCRIPTION
**Pull request by mohamadhase**
implement read_n_from_file function that takes number of messages and the path of the file to read from then return the last 5 messages in the file as pandas dataframe

### CODE REVIEW

**Review by ramahasiba — APPROVED** [review]

**Review by eleenkmail — APPROVED** [review]
Done

File: Road/source/features/feature_selected.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    month =date.dt.month
+    day =date.dt.day
+    hour =date.dt.hour
+    minute =date.dt.minute
+    second =date.dt.second
+
+    return (year, month, day, hour, minute, second)
+
+
+def extract_id(json_string: str, target_id: str) -> int:
```
Comment (zubaidasader) [inline Road/source/features/feature_selected.py]: Where will we use this function?

**Review by zubaidasader — APPROVED** [review]

File: Road/source/features/feature_selected.py (hunk 1 above)
Comment (mawada-sweis) [reply Road/source/features/feature_selected.py]: This file is not related to this PR, its related to Data Preparation PR.
The 'extract id' function is used to extract the ID from two required JSON-formatted columns which are `from_id` and `reply_to` columns.

File: Road/source/utils.py [hunk 2]
```diff
@@ -5,3 +5,35 @@ def function_name(parameter_name:type)->return_type:
     local_variable=0
     return
 """
+import pandas as pd
+
+import csv
+
+from Road import logger
```
Comment (mawada-sweis) [inline Road/source/utils.py:9]: Are you sure from the importing path?

File: Road/source/utils.py [hunk 3]
```diff
@@ -5,3 +5,35 @@ def function_name(parameter_name:type)->return_type:
     local_variable=0
     return
 """
+import pandas as pd
+
+import csv
```
Comment (mawada-sweis) [inline Road/source/utils.py]: `csv` is not used in the code, so the import statement is unnecessary.

File: Road/source/utils.py [hunk 4]
```diff
@@ -5,3 +5,35 @@ def function_name(parameter_name:type)->return_type:
...
+
```
Comment (mawada-sweis) [inline Road/source/utils.py]: According to Pep8 format, there should be no line space between the import statements.

File: Road/source/utils.py [hunk 5]
```diff
@@ -5,3 +5,35 @@ def function_name(parameter_name:type)->return_type:
...
+from Road import logger
+
+
+def read_n_from_file(
+    number_of_messages: int=5, file_path: str = "Road/data/interim/feature_selected.csv"
```
Comment (mawada-sweis) [inline Road/source/utils.py]: This path is not correct, Please update it.

File: Road/source/utils.py [hunk 6]
```diff
@@ -5,3 +5,35 @@ def function_name(parameter_name:type)->return_type:
...
+) -> pd.DataFrame:
+    """
+    Reads n messages from the end of a CSV file and returns them as a pandas DataFrame.
+
+    Args:
+        number_of_messages (int): The number of messages to read from the file.
```
Comment (mawada-sweis) [inline Road/source/utils.py]: `number_of_message` is optional arg

File: Road/source/utils.py [hunk 7]
```diff
@@ -5,3 +5,35 @@ def function_name(parameter_name:type)->return_type:
...
+    - FileNotFoundError: If the file specified in file_path cannot be found.
+    """
+    try:
+        data = pd.read_csv(file_path)
+        last_n_rows = data.tail(number_of_messages)
+    except FileNotFoundError as e:
+        logger.exception(f"Error reading file {file_path}: {e}")
+        raise
+    except KeyError:
+        pass  # skip the error if Unnamed: 0 not found in the dataframe columns
```
Comment (mawada-sweis) [inline Road/source/utils.py]: What is the purpose of this exception?

File: Road/endpoints/endpoints.py [hunk 8]
```diff
@@ -10,8 +11,12 @@ def _health_check(request: Request) -> dict:
     response = {
         "message": HTTPStatus.OK.phrase,
         "status-code": HTTPStatus.OK,
-        "data": {}
+        "data": {},
     }
+    print(read_n_from_file())
```
Comment (mawada-sweis) [inline Road/endpoints/endpoints.py]: Please save the returned value instead of print the result

Comment (mawada-sweis) [inline Road/endpoints/endpoints.py]: We have to read the data from `raw` file, please update the arg value

**Review by mawada-sweis — CHANGES_REQUESTED** [review]
Please follow the pep8 format.

**Comment by mawada-sweis** [comment]
Please provide a description for the PR.

File: Road/source/utils.py (hunk 2 above)
Comment (mohamadhase) [reply Road/source/utils.py:9]: i didn't understand , can you be more specific?

File: Road/source/utils.py (hunk 3 above)
Comment (mohamadhase) [reply Road/source/utils.py]: yes you are right , its from old code i wrote and deleted it

File: Road/source/utils.py (hunk 4 above)
Comment (mohamadhase) [reply Road/source/utils.py]: ok

File: Road/source/utils.py (hunk 5 above)
Comment (mohamadhase) [reply Road/source/utils.py]: what the path sohuld be? raw data?

File: Road/source/utils.py (hunk 6 above)
Comment (mohamadhase) [reply Road/source/utils.py]: i will add the default value to the comment

File: Road/source/utils.py (hunk 7 above)
Comment (mohamadhase) [reply Road/source/utils.py]: it should be removed ,  its from the old code i wrote and know there's know point of being there

File: Road/endpoints/endpoints.py (hunk 8 above)
Comment (mohamadhase) [reply Road/endpoints/endpoints.py]: this line should be deleted , it was for testing only

File: Road/source/utils.py (hunk 5 above)
Comment (mawada-sweis) [reply Road/source/utils.py]: yes, the default path should be the raw data

File: Road/source/utils.py (hunk 2 above)
Comment (mawada-sweis) [reply Road/source/utils.py:9]: never mind, I ran the file instead of app

File: Road/endpoints/endpoints.py [hunk 9]
```diff
@@ -1,6 +1,7 @@
 from http import HTTPStatus
 from fastapi import Request
 from Road import app, logger
+from Road.source.utils import read_n_from_file
```
Comment (mawada-sweis) [inline Road/endpoints/endpoints.py:4]: You should call the function and save the returned value to pass it to processed functions.

**Review by mawada-sweis — APPROVED** [review]

File: Road/source/utils.py [hunk 10]
```diff
@@ -16,7 +16,7 @@ def read_n_from_file(
     Reads n messages from the end of a CSV file and returns them as a pandas DataFrame.
 
     Args:
-        number_of_messages (int): The number of messages to read from the file. Defaults to 5
+        number_of_messages (intو optional): The number of messages to read from the file. Defaults to 5
```
Comment (Reema-Khaseeb) [inline Road/source/utils.py]: و ❌
, ✅

**Review by Reema-Khaseeb — CHANGES_REQUESTED** [review]

File: Road/source/features/data_cleaning.py [hunk 11]
```diff
@@ -1,20 +1,20 @@
...
 target = os.path.abspath(__file__)
-while(target.split("\\")[-1]!="Road"):
+while(target.split("\\")[-1] != "Road"):
     target = os.path.dirname(target)
-sys.path.append(target) 
+sys.path.append(target)
 
 # For arabic diacritics
-# pip install pyarabic 
+# pip install pyarabic
```
Comment (Reema-Khaseeb) [inline Road/source/features/data_cleaning.py:17]: No need for this comment

**Review by Dana-Ismail — APPROVED** [review]

File: Road/endpoints/endpoints.py (hunk 9 above)
Comment (Reema-Khaseeb) [inline Road/endpoints/endpoints.py]: Unused

**Review by Reema-Khaseeb — CHANGES_REQUESTED** [review]

File: Road/source/features/data_cleaning.py (hunk 11 above)
Comment (mohamadhase) [reply Road/source/features/data_cleaning.py:17]: that's not my code , you can see the data cleaning PR for this comment

File: Road/endpoints/endpoints.py (hunk 9 above)
Comment (mohamadhase) [reply Road/endpoints/endpoints.py]: i will delete this line

File: Road/source/utils.py (hunk 10 above)
Comment (mohamadhase) [reply Road/source/utils.py]: aaaaaaaaaaaaaaaaaaaaaaaaaaaaaah
done 😢

**Review by Reema-Khaseeb — APPROVED** [review]

**Pull request merged**